### PR TITLE
test: Properly clear storage dirty flag

### PR DESCRIPTION
**Pull request by chfast**
Clear storage dirty flag between transactions in EVMHost. The EIP-2200 dirty flag should be persisted only in a single transaction.

Before a storage slot could have been incorrectly considered "dirtly" and the gas cost of a `SSTORE` was lower than it should have been.

### CODE REVIEW

**Comment by axic** [comment]
This is part of the fix/investigation of #13532.

File: test/EVMHost.cpp [hunk 1]
```diff
@@ -174,10 +174,13 @@ void EVMHost::newTransactionFrame()
...
 	// Clear EIP-2929 account access indicator
 	recorded_account_accesses.clear();
-	// Clear EIP-2929 storage access indicator
+
 	for (auto& [address, account]: accounts)
 		for (auto& [slot, value]: account.storage)
-			value.access_status = EVMC_ACCESS_COLD;
+		{
+			value.access_status = EVMC_ACCESS_COLD; // Clear EIP-2929 storage access indicator
+			value.dirty = false;					// Clear EIP-2200 dirty slot flag
```
Comment (axic) [inline test/EVMHost.cpp:182]: See https://github.com/ethereum/evmc/blob/v8.0.0/include/evmc/mocked_host.hpp#L24

Comment (axic) [reply test/EVMHost.cpp:182]: This has been there since the beginning and we probably never supported it properly.

**Review by axic — DISMISSED** [review]

**Comment by axic** [comment]
@chfast need to run the tests with `--optimize` again to populate those values.

File: test/libsolidity/semanticTests/array/dynamic_array_cleanup.sol [hunk 2]
```diff
@@ -19,6 +19,7 @@ contract c {
 // gas legacyOptimized: 517027
 // storageEmpty -> 0
 // halfClear() ->
+// gas legacy: 113257
```
Comment (chfast) [inline test/libsolidity/semanticTests/array/dynamic_array_cleanup.sol:23]: why is this line added?

Comment (axic) [reply test/libsolidity/semanticTests/array/dynamic_array_cleanup.sol:23]: Gas figures are included above a threshold (~100k). With the change it is above the threshold.

**Review by ekpyron — APPROVED** [review]
Good to merge if there's no CI failures left!

**Comment by axic** [comment]
@ekpyron 🙏
